### PR TITLE
Update ListingTable to use StatisticsConverter, remove redundant statistics extraction code

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/10923

## Rationale for this change

The code that extracted statistics for ListingTable with parquet files:
1. Was incomplete and did not handle all types (like strings)
3. Did the same thing (extracted Row Group statistics from parquet files) but used different code than the ParquetExec
4. Was inefficient (accumulated by creating single row record batches for each row group)

I also personally also found the previous code somewhat hard to follow

See https://github.com/apache/datafusion/issues/10923 for more details

## What changes are included in this PR?

Update Parquet listing table to use the new `StatisticsConverter` API to create DataFusion statistics

## Are these changes tested?
Yes, they are covered by existing tests 

## Are there any user-facing changes?
ListingFiles will now have more accurate statistics

### CODE REVIEW

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 1]
```diff
@@ -295,86 +297,6 @@ impl FileFormat for ParquetFormat {
...
-            if let Some(min_value) = &mut min_values[i] {
-                min_value
-                    .update_batch(&[Arc::new(Int32Array::from_value(*s.min(), 1))])
-                    .unwrap_or_else(|_| min_values[i] = None);
-            }
-        }
-        ParquetStatistics::Int64(s) if DataType::Int64 == *fields[i].data_type() => {
-            if let Some(max_value) = &mut max_values[i] {
-                max_value
-                    .update_batch(&[Arc::new(Int64Array::from_value(*s.max(), 1))])
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/parquet.rs:338]: note how update_batch is called by first creating a single row array

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 2]
```diff
@@ -482,73 +404,101 @@ pub async fn statistics_from_parquet_meta(
...
-                        min_values[table_idx] = None;
-                    }
-                } else {
-                    *null_cnt = null_cnt.add(&Precision::Exact(num_rows as usize));
-                }
-            }
-        }
-    }
+        let file_field = file_schema.field(file_idx);
+        let Some(converter) = StatisticsConverter::try_new(
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/parquet.rs:437]: this code now uses the well tested StatisticsConverter to extract statistics from the parquet file with the correct type of array

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 3]
```diff
@@ -1449,8 +1400,15 @@ mod tests {
         // column c1
         let c1_stats = &stats.column_statistics[0];
         assert_eq!(c1_stats.null_count, Precision::Exact(1));
-        assert_eq!(c1_stats.max_value, Precision::Absent);
-        assert_eq!(c1_stats.min_value, Precision::Absent);
+        // Note in ASCII lower case is greater than upper case
+        assert_eq!(
+            c1_stats.max_value,
+            Precision::Exact(ScalarValue::from("bar"))
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/parquet.rs:1406]: strings were previously not handled

File: datafusion/sqllogictest/test_files/explain.slt [hunk 4]
```diff
@@ -287,20 +287,20 @@ query TT
...
+02)--ParquetExec: file_groups={1 group: [[WORKSPACE_ROOT/parquet-testing/data/alltypes_plain.parquet]]}, projection=[id, bool_col, tinyint_col, smallint_col, int_col, bigint_col, float_col, double_col, date_string_col, string_col, timestamp_col], limit=10, statistics=[Rows=Exact(8), Bytes=Absent, [(Col[0]: Min=Exact(Int32(NULL)) Max=Exact(Int32(NULL)) Null=Exact(0)),(Col[1]: Min=Exact(Boolean(NULL)) Max=Exact(Boolean(NULL)) Null=Exact(0)),(Col[2]: Min=Exact(Int32(NULL)) Max=Exact(Int32(NULL)) Null=Exact(0)),(Col[3]: Min=Exact(Int32(NULL)) Max=Exact(Int32(NULL)) Null=Exact(0)),(Col[4]: Min=Exact(Int32(NULL)) Max=Exact(Int32(NULL)) Null=Exact(0)),(Col[5]: Min=Exact(Int64(NULL)) Max=Exact(Int64(NULL)) Null=Exact(0)),(Col[6]: Min=Exact(Float32(NULL)) Max=Exact(Float32(NULL)) Null=Exact(0)),(Col[7]: Min=Exact(Float64(NULL)) Max=Exact(Float64(NULL)) Null=Exact(0)),(Col[8]: Min=Exact(Binary(NULL)) Max=Exact(Binary(NULL)) Null=Exact(0)),(Col[9]: Min=Exact(Binary(NULL)) Max=Exact(Binary(NULL)) Null=Exact(0)),(Col[10]: Min=Exact(TimestampNanosecond(NULL, None)) Max=Exact(TimestampNanosecond(NULL, None)) Null=Exact(0))]]
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/explain.slt:291]: I don't fully understand why these statistics are all reported as null / None now. -- I need to investagate

File: datafusion/sqllogictest/test_files/explain.slt [hunk 5]
```diff
@@ -287,20 +287,20 @@ query TT
...
+01)GlobalLimitExec: skip=0, fetch=10, statistics=[Rows=Exact(8), Bytes=Absent, [(Col[0]: Null=Exact(0)),(Col[1]: Null=Exact(0)),(Col[2]: Null=Exact(0)),(Col[3]: Null=Exact(0)),(Col[4]: Null=Exact(0)),(Col[5]: Null=Exact(0)),(Col[6]: Null=Exact(0)),(Col[7]: Null=Exact(0)),(Col[8]: Null=Exact(0)),(Col[9]: Null=Exact(0)),(Col[10]: Null=Exact(0))]]
+02)--ParquetExec: file_groups={1 group: [[WORKSPACE_ROOT/parquet-testing/data/alltypes_plain.parquet]]}, projection=[id, bool_col, tinyint_col, smallint_col, int_col, bigint_col, float_col, double_col, date_string_col, string_col, timestamp_col], limit=10, statistics=[Rows=Exact(8), Bytes=Absent, [(Col[0]: Null=Exact(0)),(Col[1]: Null=Exact(0)),(Col[2]: Null=Exact(0)),(Col[3]: Null=Exact(0)),(Col[4]: Null=Exact(0)),(Col[5]: Null=Exact(0)),(Col[6]: Null=Exact(0)),(Col[7]: Null=Exact(0)),(Col[8]: Null=Exact(0)),(Col[9]: Null=Exact(0)),(Col[10]: Null=Exact(0))]]
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/explain.slt:291]: this is more correct -- these columns have no nulls:

```shell

There are no nulls in this file:
andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion$ datafusion-cli -c 'select * from "parquet-testing/data/alltypes_plain.parquet"';
DataFusion CLI v39.0.0
+----+----------+-------------+--------------+---------+------------+-----------+------------+------------------+------------+---------------------+
| id | bool_col | tinyint_col | smallint_col | int_col | bigint_col | float_col | double_col | date_string_col  | string_col | timestamp_col       |
+----+----------+-------------+--------------+---------+------------+-----------+------------+------------------+------------+---------------------+
| 4  | true     | 0           | 0            | 0       | 0          | 0.0       | 0.0        | 30332f30312f3039 | 30         | 2009-03-01T00:00:00 |
| 5  | false    | 1           | 1            | 1       | 10         | 1.1       | 10.1       | 30332f30312f3039 | 31         | 2009-03-01T00:01:00 |
| 6  | true     | 0           | 0            | 0       | 0          | 0.0       | 0.0        | 30342f30312f3039 | 30         | 2009-04-01T00:00:00 |
| 7  | false    | 1           | 1            | 1       | 10         | 1.1       | 10.1       | 30342f30312f3039 | 31         | 2009-04-01T00:01:00 |
| 2  | true     | 0           | 0            | 0       | 0          | 0.0       | 0.0        | 30322f30312f3039 | 30         | 2009-02-01T00:00:00 |
| 3  | false    | 1           | 1            | 1       | 10         | 1.1       | 10.1       | 30322f30312f3039 | 31         | 2009-02-01T00:01:00 |
| 0  | true     | 0           | 0            | 0       | 0          | 0.0       | 0.0        | 30312f30312f3039 | 30         | 2009-01-01T00:00:00 |
| 1  | false    | 1           | 1            | 1       | 10         | 1.1       | 10.1       | 30312f30312f3039 | 31         | 2009-01-01T00:01:00 |
+----+----------+-------------+--------------+---------+------------+-----------+------------+------------------+------------+---------------------+
8 row(s) fetched.
Elapsed 0.009 seconds.
```

**Review by xinlifoobar — APPROVED** [review]

**Comment by alamb** [comment]
Thank you for the review @xinlifoobar 

@Ted-Jiang  I wonder if you might be interested in reviewing this PR as well?
